### PR TITLE
[Draft] refactor bucket batch sampler to support DDP sampling

### PR DESCRIPTION
**Pull request by guoqing-zhou**
(**NOTE:** _**delete** these instructional lines as you fill-out this PR template_)

(**NOTE:** _template is designed to be filled-in and used as the **squashed commit message for the entire PR**. _Italicized text_ is intended to be deleted as you fill in this template. Use the text between the `---`)

---

_High level summary of changes. Try to keep this as short and informative as possible: less is more._

_Describe your changes. You can be more detailed and descriptive here. If it is a code change, Be sure to answer:_
  - _What is changing?_
  - _What is the new or fixed functionality?_
  - _Why or when would someone want to use these changes?_
  - _How can someone use these changes?_
---

## Summary
_High level summary of changes. Try to keep this as short and informative as possible: less is more._

## Details
_Describe your changes. You can be more detailed and descriptive here._

## Usage
_How does a user interact with the changed code?_
```python
python -m your.new.module -and -all -options
```

## Testing
_How do you prove that your code behaves the way you claim?_

Tests for these changes can be run via:
```shell
pytest -v tests/your/new/or/existing/test_functions.py::test_function
```

(**NOTE:** _also **delete** this checklist as you fill-out this PR template_)

**Most of the changes** to files with extensions `*.py`, `*.yaml`, `*.yml`, `Dockerfile*` or `requirements.txt` **DO REQUIRE both `pytest-` and `jet-` CI stages**.

- [ ] Did you review the [Before your PR is "Ready for review" section](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#before-pr-ready) before asking for review?
- [ ] Did you make sure your changes have tests? Did you test your changes locally?
- [ ] Can you add [the `SKIP_CI` label](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#skip-ci) to your PR?
- [ ] Can you add [the `PYTEST_NOT_REQUIRED` label](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#skip-pytest) to your PR?
- [ ] Can you add [the `JET_NOT_REQUIRED` label](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#skip-jet) to your PR?

### CODE REVIEW

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 1]
```diff
@@ -486,50 +490,97 @@ def __init__(
...
+            element_bucket_indices = self.element_bucket_indices[indices]
+            # element indices reordered for each bucket
+            reordered_element_indices = indices[torch.argsort(element_bucket_indices, stable=True)]
 
         # bucket sizes, including the buckets for < bucket_boundaries[0] and >= bucket_boundaries[-1]
-        bucket_sizes = torch.bincount(element_bucket_indices, minlength=len(bucket_boundaries) + 1)
+        bucket_sizes = torch.bincount(element_bucket_indices, minlength=len(self.bucket_boundaries) + 1)
 
         # bucket segments
         bucket_segments = torch.cumsum(bucket_sizes, dim=0)[:-1]
```
Comment (DejunL) [inline sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py:566]: could we rename `bucket_segments` to bucket_sizes_cumsum` or something like that to indicate it's about how many elements there are in each and cumsummed

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 2]
```diff
@@ -486,50 +490,97 @@ def __init__(
...
-            self.bucket_element_indices.append(
-                reordered_element_indices[bucket_segments[bucket_idx] : bucket_segments[bucket_idx + 1]].tolist()
-            )
+            self.bucket_element_indices[bucket_idx][:] = reordered_element_indices[
+                bucket_segments[bucket_idx] : bucket_segments[bucket_idx + 1]
+            ].tolist()
+
         self.bucket_sizes = bucket_sizes[1 : (self.num_buckets + 1)]
 
         self.num_samples = torch.sum(self.bucket_sizes).item()
```
Comment (DejunL) [inline sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py:576]: couldn't this be either len(indices) or len(self.sizes)? i.e., I don't think we need to do sum

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 3]
```diff
@@ -486,50 +490,97 @@ def __init__(
...
         if self.num_samples == 0:
             raise RuntimeError("The sizes of all elements in the dataset are outside the bucket ranges provided")
-        if self.num_samples < len(self.sizes):
+        if self.num_samples < len(element_bucket_indices):
```
Comment (DejunL) [inline sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py:579]: when would this be triggered?

**Review by DejunL — CHANGES_REQUESTED** [review]
Did my first pass. Have some minor comments. I would need to look more into the how lightning engage with the DistributedSampler to get an understanding of the underlying logics so to come up with a test

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py (hunk 3 above)
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py:579]: if there are elements outside the boundaries, maybe shouldn't be triggered here.

Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py]: move to init, and just compute the number of outliers there

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 4]
```diff
@@ -390,12 +392,14 @@ def __init__(
                 Default to  {}.
             shuffle: A boolean indicating whether to shuffle the dataset and buckets. Defaults to True.
             generator: Generator used in sampling. Defaults to None.
+            sampler: Sampler to get element indices for this batch sampler, required for distributed sampling. Defaults to None.
```
[the remark this answers is not in the document]
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py]: updated

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 5]
```diff
@@ -390,12 +392,14 @@ def __init__(
...
+            num_batches: Number of mini-batches, required for distributed sampling. Defaults to None.
```
[the remark this answers is not in the document]
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py]: updated

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 6]
```diff
@@ -486,50 +490,97 @@ def __init__(
...
         self.bucket_sizes: torch.Tensor  # number of elements in each bucket
         self.bucket_element_indices: List[List[int]]  # List of elements' indices for each bucket
 
         # bucket index for each element
-        element_bucket_indices = torch.bucketize(sizes, bucket_boundaries, right=True)
+        self.element_bucket_indices = torch.bucketize(sizes, self.bucket_boundaries, right=True)
+
+        if sampler is not None and not isinstance(sampler, Sampler):  # type: ignore
+            raise TypeError(
+                f"sampler should be a sampler class inherited from torch.utils.data.Sampler, but got sampler={sampler}"
```
[the remark this answers is not in the document]
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py]: updated

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 7]
```diff
@@ -486,50 +490,97 @@ def __init__(
...
+                "generator": self.generator,
+                "sampler": self.sampler,
+                "num_batches": self.num_batches,
+            },
+        )
+        setattr(self, "__pl_saved_default_kwargs", {})
+        setattr(self, "__pl_saved_arg_names", ())
+        return self
+
+    def _compute_bucket_element_indices(self, indices: Optional[List[int]] = None):
```
[the remark this answers is not in the document]
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py:561]: updated

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py (hunk 1 above)
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py]: updated

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py (hunk 2 above)
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py]: There are potentially elements outside the bucket boundaries, and they are filtered, so still need to compute sum.

File: sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py [hunk 8]
```diff
@@ -486,50 +490,97 @@ def __init__(
...
+
+        if num_batches is not None and (not isinstance(num_batches, int) or num_batches <= 0):
+            raise ValueError(f"num_batches should be a positive integer, but got num_batches = {num_batches}")
+
+        self.num_batches = num_batches
+
+    def setup_for_pytorch_lightning(self) -> Sampler:
+        """Add additional attributes necessary to be compatible with pytorch lightning distributed sampling.
+
+        When used for distributed sampling, Pytorch Lightning will use these attributes to re-init this batch sampler and
```
[the remark this answers is not in the document]
Comment (guoqing-zhou) [reply sub-packages/bionemo-size-aware-batching/src/bionemo/size_aware_batching/sampler.py:534]: updated
